livy/src/main/java/org/apache/zeppelin/livy/BaseLivyInterprereter.java

```diff
@@ -140,4 +158,304 @@ public FormType getFormType() {
+      SessionInfo sessionInfo = SessionInfo.fromJson(
+          callRestAPI("/sessions", "POST", request.toJson()));
+      long start = System.currentTimeMillis();
+      // pull the session status until it is idle or timeout
+      while (!sessionInfo.isReady()) {
+        LOGGER.info("Session {} is in state {}, appId {}", sessionInfo.id, sessionInfo.state,
+            sessionInfo.appId);
+        if (sessionInfo.isFinished()) {
+          String msg = "Session " + sessionInfo.id + " is finished, appId: " + sessionInfo.appId
+              + ", log: " + sessionInfo.log;
+          LOGGER.error(msg);
+          throw new LivyException(msg);
+        }
+        if ((System.currentTimeMillis() - start) / 1000 > sessionCreationTimeout) {
+          String msg = "The creation of session " + sessionInfo.id + " is timeout within "
+              + sessionCreationTimeout + " seconds, appId: " + sessionInfo.appId
+              + ", log: " + sessionInfo.log;
+          LOGGER.error(msg);
+          throw new LivyException(msg);
+        }
+        Thread.sleep(1000);
+        sessionInfo = getSessionInfo(sessionInfo.id);
+      }
+      return sessionInfo;
+    } catch (Exception e) {
+      LOGGER.error("Error when creating livy session for user " + user, e);
+      throw new LivyException(e);
+    }
+  }
+
+  private SessionInfo getSessionInfo(int sessionId) throws LivyException {
+    return SessionInfo.fromJson(callRestAPI("/sessions/" + sessionId, "GET"));
+  }
+
+  public InterpreterResult interpret(String code, boolean displayAppInfo)
+      throws LivyException {
+    StatementInfo stmtInfo = executeStatement(new ExecuteRequest(code));
+    // pull the statement status
+    while (!stmtInfo.isAvailable()) {
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        throw new LivyException(e);
+      }
+      stmtInfo = getStatementInfo(stmtInfo.id);
+    }
+    return getResultFromStatementInfo(stmtInfo, displayAppInfo);
+  }
+
+  private InterpreterResult getResultFromStatementInfo(StatementInfo stmtInfo,
+                                                       boolean displayAppInfo) {
+    if (stmtInfo.output.isError()) {
+      return new InterpreterResult(InterpreterResult.Code.ERROR, stmtInfo.output.evalue);
+    } else {
+      //TODO(zjffdu) support other types of data (like json, image and etc)
+      String result = stmtInfo.output.data.plain_text;
+      if (result != null) {
+        result = result.trim();
+        if (result.startsWith("<link")
+            || result.startsWith("<script")
+            || result.startsWith("<style")
+            || result.startsWith("<div")) {
+          result = "%html " + result;
+        }
+      }
+      if (displayAppInfo) {
+        //TODO(zjffdu), use multiple InterpreterResult to display appInfo
+        StringBuilder outputBuilder = new StringBuilder();
+        outputBuilder.append("%angular ");
```

Comment:
This is from #1426, I will change it to multiple InterpreterResult in a followup PR.